delivery-service/ds-lib/src/lib.rs

```diff
@@ -0,0 +1,248 @@
+//! # OpenMLS Delivery Service Library
+//!
+//! This library provides structs and necessary implementations to interact with
+//! the OpenMLS DS.
+//!
+//! Clients are represented by the `ClientInfo` struct.
+
+use openmls::prelude::*;
+
+/// Information about a client.
+/// To register a new client create a new `ClientInfo` and send it to
+/// `/clients/register`.
+#[derive(Debug, Default, Clone, PartialEq)]
+pub struct ClientInfo {
+    pub client_name: String,
+    pub key_packages: ClientKeyPackages,
+    pub id: Vec<u8>,
+    pub msgs: Vec<MLSMessage>,
+    pub welcome_queue: Vec<Welcome>,
+}
+
+/// The DS returns a list of key packages for a client as `ClientKeyPackages`.
+/// This is a tuple struct holding a vector of `(Vec<u8>, KeyPackage)` tuples,
+/// where the first value is the key package hash (output of `KeyPackage::hash`)
+/// and the second value is the corresponding key package.
+#[derive(Debug, Default, Clone, PartialEq)]
+pub struct ClientKeyPackages(pub Vec<(Vec<u8>, KeyPackage)>);
+
+impl ClientInfo {
+    /// Create a new `ClientInfo` struct for a given client name and vector of
+    /// key packages with corresponding hashes.
+    pub fn new(client_name: String, key_packages: Vec<(Vec<u8>, KeyPackage)>) -> Self {
+        Self {
+            client_name,
+            id: key_packages[0].1.credential().identity().to_vec(),
+            key_packages: ClientKeyPackages(key_packages),
+            msgs: Vec::new(),
+            welcome_queue: Vec::new(),
+        }
+    }
+
+    /// The identity of a client is defined as the identity of the first key
+    /// package right now.
+    pub fn id(&self) -> &[u8] {
+        self.key_packages.0[0].1.credential().identity()
+    }
+}
+
+/// The DS returns a list of messages on `/recv/{name}`, which is a `Vec<Message>`.
+/// A `Message` is either an `MLSMessage` or a `Welcome` message (see OpenMLS)
+/// for details.
+#[derive(Debug, Clone, PartialEq)]
+pub enum Message {
+    /// An `MLSMessage` is either an OpenMLS `MLSCiphertext` or `MLSPlaintext`.
+    MLSMessage(MLSMessage),
+
+    /// An OpenMLS `Welcome` message.
+    Welcome(Welcome),
+}
+
+/// A generalisation of the `MLSCiphertext` and `MLSPlaintext` messages from
+/// OpenMLS.
+#[allow(clippy::large_enum_variant)]
+#[derive(Debug, Clone, PartialEq)]
+pub enum MLSMessage {
+    /// An OpenMLS `MLSCiphertext`.
+    MLSCiphertext(MLSCiphertext),
+
+    /// An OpenMLS `MLSPlaintext`.
+    MLSPlaintext(MLSPlaintext),
+}
+
+impl MLSMessage {
+    /// Get the group ID as plain byte vector.
+    pub fn group_id(&self) -> Vec<u8> {
+        match self {
+            MLSMessage::MLSCiphertext(m) => m.group_id.as_slice(),
+            MLSMessage::MLSPlaintext(m) => m.group_id().as_slice(),
```

Comment:
Is there a particular reason, that we access the ciphertext by field and the plaintext by getter? Same for the `epoch` below.

Reply:
Because that's how it currently is in the library... the API needs to be cleaned up.